### ISSUE TITLE
Multihead Attention Container GPU Memory Leak

### ISSUE THREAD

**Issue by knagrecha** [pytorch/text#1270]
## 🐛 Bug
**Describe the bug**
The Multihead Attention Container suffers a memory leak. This leak ONLY occurs on the first use of the module.

**To Reproduce**
Pass a tensor through the module. Delete the tensor. Check memory usage before and after.

Sample code:
```
from torch.nn import Linear
from torchtext.nn import MultiheadAttentionContainer, InProjContainer, ScaledDotProduct
import torch
from pynvml import nvmlInit, nvmlDeviceGetHandleByIndex, nvmlDeviceGetMemoryInfo

def get_free_space(idx=0):
    nvmlInit()
    h = nvmlDeviceGetHandleByIndex(idx)
    info = nvmlDeviceGetMemoryInfo(h)
    return info.free

in_proj_container = InProjContainer(Linear(768, 768),
                                            Linear(768, 768),
                                            Linear(768, 768))
        
mha = MultiheadAttentionContainer(16, in_proj_container,
                                       ScaledDotProduct(), Linear(768, 768)).to("cuda:0")


with torch.no_grad():
    print(get_free_space(0))
    a_detach = torch.zeros((128, 129, 768)).to("cuda:0")
    print(get_free_space(0))
    print("***********")
    b = mha(a_detach, a_detach, a_detach, None)
    print("***********")
    
    del b
    del a_detach
    gc.collect()
    torch.cuda.empty_cache()
    print(get_free_space(0))
    
    
    a_detach = torch.zeros((128, 129, 768)).to("cuda:0")
    print(get_free_space(0))
    print("***********")
    b = mha(a_detach, a_detach, a_detach, None)
    print("***********")
    print(get_free_space(0))
    del b
    del a_detach
    
    gc.collect()
    torch.cuda.empty_cache()
    print(get_free_space(0))
```

**Expected behavior**
Ideally, memory prior to pass and post-pass should have been the same. I even used torch.no_grad() to prevent gradient accumulation. Instead, I found that a rather large amount of memory was lost!

**Screenshots**
![image](https://user-images.githubusercontent.com/32966638/113199975-a29efc80-921c-11eb-9e8d-eb1a633808a8.png)

Notice that the second time, all memory is recovered to what it was prior to the pass. But in the first pass, roughly 50MB is lost.

**Environment**

 - PyTorch Version: 1.8.1
 - OS: Linux
 - How you installed PyTorch: pip
 - Python version: 3.8
 - CUDA/cuDNN version: 11.1
 - GPU models and configuration: Two GTX 1080s. Only using one for this issue though.

**Context**
The get_free_space function is just a wrapper around pynvml, so the readings should be accurate.

**Comment by knagrecha** [comment]
Misread my own logs...not an issue.

**Comment by knagrecha** [comment]
I've been digging in further, I think the issue comes from the InProjContainer class (also defined in torchtext.nn.multiheadattentioncontainer.py. Will try and find out what's going on.

**Comment by knagrecha** [comment]
It seems really strange, but it apparently is coming from the nn.Linear I sent to inprojcontainer. See this new code snippet. I'll post this on the general PyTorch forum rather than torchtext specifically.

`linear_layer = torch.nn.Linear(768, 768).to("cuda:0")
with torch.no_grad():
    print(utilities.get_free_space(0))
    a_detach = torch.zeros((128, 129, 768)).to("cuda:0")
    d = linear_layer(a_detach)
    del d
    del a_detach

    gc.collect()
    torch.cuda.empty_cache()
    print(utilities.get_free_space(0))
    `
![image](https://user-images.githubusercontent.com/32966638/113336190-d93b4c80-92da-11eb-8e2f-dad56ba359cb.png)
